### PR TITLE
Support Flow this parameter annotations

### PR DESCRIPTION
**Pull request by dsainati1**
This PR adds support for Flow `this` parameter annotations. 

E.g.

```
function foo (this : number) {}
``` 

- [X] I’ve added tests to confirm my change works.
- [X] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [X] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
**Prettier pr-9457**
[Playground link](https://deploy-preview-9457--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAKtgXmwAoYALASwGclsAjCCAGzgEMoAabAHSm34GDWtKAFcAtnTgAnLr0EKefRQLoiJU2UpX8AlIQB89Ri3YBuEBxAQMMCtCrJQradIgB3AAouEjlKyZ3VjRHKzppVjAAazgYAGUMSIooAHNkGGlROCs4STgAE3yCgBl2FNFWFLgAMQhpcVYYO1TkEFZRGAhLEDIYcSYAdUp4KkSwODjfCjsAN2m0VrAqUJBkqhkYTwiUhuQAMwD1qwArKgAPACEI6Ni41nE4YuS4fcPskFOzuOSUlgBFUQQeCvJhHECJaTraStPZMDzdDDSZIwAYUfLkZAADgADFZERB1gMIhhWoi4FCZi8rABHQHwLY2PxtKgAWigcAKBW60jgtIoPK2lV2SAOoPe63EFHSmXFP3+dJeIreVhgrDoqPRZGQACYVREKEwfgBhCDiYUgckAVm6onWuDVflFYJmWQAklAirA4mAkbYAIIeuLoFgg9YAXzDQA)
```sh
--parser flow
```

**Input:**
```jsx
type T = (this: boolean, 
          a: number, 
          
          b: number, 
         ) => boolean;
```

**Output:**
```jsx
type T = (
  this:
  boolean,
  a: number,

  b: number
) => boolean;

```

**Comment by fisker** [comment]
Do you think we  should treat it as a parameter?

**Prettier pr-9457**
[Playground link](https://deploy-preview-9457--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAgtgXmwAoAdKbbGACwEsBnJbAZQFcMsAnAGWgHNeUAf0HChoiSKnj+AFUxwANNnKU16lRWwAjJlFYBbbXE7LyASkIA+HRAgAbOAEMoAbnLl0WbACFCJVWwAfRoGJjYOE0kZGLE46PjpPnksMy0NNUDdbH0jEzTLAhttO0cXDygQRRAIDBhaaHpkUCdOTggAdwAFVoQmlCd7Dqc0JurtTicwAGs4GGYMKdohZBhOViUQODyAEx24He4XPlYnPjgAMQhOAycYepWB1hgIKpBqGAN7AHU6eHpFmA4Mw+rR6gA3MFoZDgehjEDLegmGBdSZ8W7IABmgyR1QAVvQAB4+SYzObMJwGODcZZwLE4zYEwnMZZ8RwARVYEHg9PsuJAi04SM4MMx9k6bwwnGWMG+tB2NGQAA4AAzVKUQJHfSYYGFSuDC8F06oARy58FRtX6ICc9AAtFA4AcDm9OHAzbQ3aizhikNi+ZskQZaKt1oHWRzzXS-QzqjAnNo5QrqMgAExxya0eysgDCEAMvq29AArG9WEjZAn+v7+eCNgBJKD7WDMMDSuq4JvMdCOXlIgC+-aAA)
```sh
--parser flow
```

**Input:**
```jsx
type A = (
  this: SupperLongLongLongLongLongLongLongLongLongLongLongType, 
          
  b: number, 
) => boolean;

type B = (
  _this: SupperLongLongLongLongLongLongLongLongLongLongLongType, 
          
  b: number, 
) => boolean


```

**Output:**
```jsx
type A = (
  this: SupperLongLongLongLongLongLongLongLongLongLongLongType,
  b: number
) => boolean;

type B = (
  _this: SupperLongLongLongLongLongLongLongLongLongLongLongType,

  b: number
) => boolean;

```

**Comment by dsainati1** [comment]
I am ambivalent about it; the `this` parameter is not a true parameter in the JavaScript sense.

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -3913,7 +3913,11 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
     parts.push("this: ");
     parts.push(path.call(print, "this"));
     if (printed.length > 0 || fun.rest) {
-      parts.push(",", line);
+      if (isNextLineEmpty(options.originalText, fun.this, options.locEnd)) {
```
Comment (evilebottnawi) [inline src/language-js/printer-estree.js:3916]: I think this is really bad idea

Comment (dsainati1) [reply src/language-js/printer-estree.js:3916]: I just took this from how parameters are handled above. What specifically about it do you think is bad?

Comment (evilebottnawi) [reply src/language-js/printer-estree.js:3916]: We should avoid to using original source code

Comment (dsainati1) [reply src/language-js/printer-estree.js:3916]: In that case what would you suggest instead?

**Comment by thorn0** [comment]
One more inconsistency:

**Prettier pr-9457**
[Playground link](https://deploy-preview-9457--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAgtgXmwAoAdKbbGACwEsBnJbAH2wGUBXDLAJwBloAcwFRhQkWNETp42VKEAVTDlYAxCBABCAQx46AXgJ5wAtgEYATAGYALAFYANNnKUARkygcTruDyfkASkIAPmxXDQAbOG0oAG5ycnQsbE1CEhdsAH0aBiZWTm5fGXkSyTLi8rlBJWS1DR09bUMIY3Nre38KMI8vHz9nKCCCUPCIKJiEqBAHEAgMGFpoemRQXR4IAHcABV0EZZRtCI3tNGWZ1x5tMABrOBg2DCvaUWQYHg44GdMfABMfuB+fBigg42kEcHUPBM2hgCxeBw4MAg0xA1BgJgiAHU6PB6I8wHA2HtaAsAG4ktDIcD0M4gZ70XwwLaXQTQ5AAM0ODJmACt6AAPTSXG53NjaExwPjPOAcrmfEB8-lsZ6CKIARQ4EHgsoi3JAjx4DJ4VPZEU2KIwPGeMExtB+NGQAA4AAwzS0QBmYy4YKmWuBG0kymYAR018GZc32IG09AAtFA4ACASjjKHaMZmWC2UhObr5QyTLRXu98yr1WGZTm5TMYNpXLb7dRkBYa5daBEVQBhCAmbMgf12FEcBkKOv7XN60kfACSUH+sDYYCt81wc7Y6CiOoZAF9t0A)
```sh
--parser flow
```

**Input:**
```jsx
type A = (
  this: | SupperLongLongLongLongLongLongLongLongLongLongLongType | FooBarBazLorem12345, 
  b: number, 
) => boolean;

type B = (
  _this: | SupperLongLongLongLongLongLongLongLongLongLongLongType | FooBarBazLorem12345, 
  b: number, 
) => boolean


```

**Output:**
```jsx
type A = (
  this: | SupperLongLongLongLongLongLongLongLongLongLongLongType
  | FooBarBazLorem12345,
  b: number
) => boolean;

type B = (
  _this:
    | SupperLongLongLongLongLongLongLongLongLongLongLongType
    | FooBarBazLorem12345,
  b: number
) => boolean;

```

File: src/language-js/printer-estree.js (hunk 1 above)
Comment (thorn0) [reply src/language-js/printer-estree.js:3916]: It's the right thing to do. Prettier preserves blank lines between parameters.

Comment (fisker) [reply src/language-js/printer-estree.js:3916]: https://github.com/prettier/prettier/pull/5836

**Comment by fisker** [comment]
This should work

```js
const printed = [
  print.cal(print, "this"),
  ...print.call(print, "params"),
];

const parts = []

// loop printed, insert comma and blank lines
```

**Comment by dsainati1** [comment]
Unfortunately the Flow parser treats the `name` field of the `this` parameter node as `null`, so it isn't as simple as just adding it to the param list.

File: src/language-js/printer-estree.js [hunk 2]
```diff
@@ -3878,35 +3878,40 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
...
-    printed = path.map((childPath, index) => {
-      const parts = [];
-      const param = childPath.getValue();
+    parts = path.map((childPath) => {
+      return { value: childPath.getValue(), param: print(childPath) };
+    }, paramsField);
+  }
 
-      parts.push(print(childPath));
+  if (fun.this) {
```
Comment (fisker) [inline src/language-js/printer-estree.js:3888]: Can we move this above the parameters print ?

File: src/language-js/printer-estree.js [hunk 3]
```diff
@@ -3878,35 +3878,40 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
...
+    fun.this.name = "this";
```
Comment (fisker) [inline src/language-js/printer-estree.js:3889]: Instead of modify the ast we can check parent in `FunctionTypeParam`

Comment (dsainati1) [reply src/language-js/printer-estree.js:3889]: the case for `FunctionTypeParam` prints a whole identifier, so re-implementing all the spacing logic there seems redundant. Assigning `this` to the ast gets us all this behavior for free.

Comment (fisker) [reply src/language-js/printer-estree.js:3883]: We are trying not  to [mutate the AST](https://github.com/prettier/prettier/issues/8122), abd that part seems simple

```js
node.name ?
path.call(print, "name") :
parent.this =node ? "this ":""
```

Should work

Comment (dsainati1) [reply src/language-js/printer-estree.js]: Yea the issue was that doing this alone causes the test case above with the multi-line union types to be different between `this` params and regular params. Adding an extra condition to the printing for union types seems to fix it though.

**Review by thorn0 — APPROVED** [review]

**Comment by fisker** [comment]
@thorn0 I changed some `parameters` check logic on other part https://github.com/prettier/prettier/pull/9457/commits/7f4e6da14191df4f247e2eb0a5d98f82050eaf56

**Review by fisker — APPROVED** [review]

File: src/language-js/printer-estree.js [hunk 4]
```diff
@@ -3870,42 +3880,44 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
...
+    ) {
+      printed.push(", ");
+    } else if (isNextLineEmpty(options.originalText, node, options.locEnd)) {
+      printed.push(",", hardline, hardline);
+    } else {
+      printed.push(",", line);
+    }
   }
 
   if (fun.rest) {
```
Comment (thorn0) [inline src/language-js/printer-estree.js:3923]: @fisker Overall, looks good. Did you consider printing `fun.rest` the same way by adding it to the `parameters` array first? And how about a `parameters` → `printedParameters` rename?

Comment (fisker) [reply src/language-js/printer-estree.js:3923]: I did think about it, but maybe later, it's not related to this PR.

**Review by thorn0 — APPROVED** [review]

**Review by evilebottnawi — APPROVED** [review]

**Comment by thorn0** [comment]
A bit late, but I found this:

**Prettier pr-9457**
[Playground link](https://deploy-preview-9457--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAjgXmwAoYALASwGcltKYAncqAcwEp8A+ANwnIBMA3CAA0ICBhjlolZKACG9ehADuABQUIZKOT34iQAI3pywAazgwAyhhNNmyBgFc4ouAFsDcPny8AZOSyOcsxwAGIQ9G5yMJIsyCByjjAQ+qQwbgA2AOoU8JQ2YHCWmuSSXKVo8WCUMqJMlHD0MKrGzFHIAGZyGQ2iAFaUAB4AQsZmFpZybnC+THCd3b0gA4OWdhlwAIqOEPALPS4gNvQN9PEdGSr6GIywWfxkyAAcAAyiNxANWcYY8Tdwpy481EuB28Ba4i0CUoAFooHAvF59PQ4KDyCiWsF2kgugdRA03OQHPRnPj1lswfMcYtDjA5AZ7nxHkgAEyiBhycgZOwAYQgbmxIABAFZ9I4GgAVelaXFLLjOACSUB8sEsYEYEgAgsrLOgNvsGgBfQ1AA)
```sh
--arrow-parens avoid
--parser flow
```

**Input:**
```jsx
type q= (this: string)=>void;
```

**Output:**
```jsx
type q = this: string => void;

```

**Second Output:**
```jsx
SyntaxError: Unexpected token `:`, expected the token `;` (1:14)
> 1 | type q = this: string => void;
    |              ^
  2 | 
```

**Comment by fisker** [comment]
Should be problem in `isFlowShorthandWithOneArg` part

**Comment by fisker** [comment]
@thorn0 After refactoring function parameters print #9489 , this get fixed https://github.com/prettier/prettier/pull/9489/commits/0ebbede9da03d3135d6886b97c85fe0f7b2673f9

**Pull request merged**